src/Razor/src/Microsoft.AspNetCore.Razor.Language/DefaultRequiredAttributeDescriptor.cs

```diff
@@ -11,14 +13,16 @@ internal class DefaultRequiredAttributeDescriptor : RequiredAttributeDescriptor
             string value,
             ValueComparisonMode valueComparison,
             string displayName,
-            RazorDiagnostic[] diagnostics)
+            RazorDiagnostic[] diagnostics,
+            Dictionary<string, string> metadata)
```

Comment:
hmm. I'd like to know more about why we need metadata here as opposed to on the top-level tag helper. If this is just for convenience, I'd suggest trying to rework the logic. This is going to result in a lot more dictionaries in memory, being serialized, etc.

Reply:
I did it this way so that in the future a single `TagHelperDescriptor` can have both directive and non-directive bound attributes. Otherwise that won't be possible.

Reply:
> Wait, i'm confused, on the required attribute descriptor? Aren't you just using the BoundAttributeDescriptor's metdata? I think I'm missing something.

Sadly, I need a way of special casing directive attributes in both. See `DefaultRequiredAttributeDescriptorBuilder.cs`